iroh-net/src/hp/derp/http/client.rs

```diff
@@ -832,6 +867,125 @@ impl Client {
         }
         Ok(())
     }
+
+    /// The tailscale code has a complicated way of dealing with the initial logging of the
+    /// `peers_present` map. Instead, we have an initial sleep of 5 seconds, and then print whatever
+    /// peers we know are present in that derp server. After that, we print the peers present every
+    /// 10 second interval.
+    ///
+    pub async fn run_mesh_client(
+        self,
+        packet_forwarder_handler: PacketForwarderHandler<Client>,
+    ) -> anyhow::Result<()> {
+        let redial_delay = Duration::from_secs(5);
+
+        // connect to the remote server & request to watching the remote's state changes
+        let ignore_key = self.public_key();
+        loop {
+            let (server_public_key, last_conn_gen) = match self.watch_connection_changes().await {
+                Ok(key) => key,
+                Err(e) => {
+                    tracing::warn!("error connecting to derp server {e}");
+                    tokio::time::sleep(redial_delay).await;
+                    continue;
+                }
+            };
+
+            if server_public_key == ignore_key {
+                bail!("detected self-connect; closing this client");
+            }
+
+            let peers_present = PeersPresent::new(server_public_key.clone());
+            tracing::info!("Connected to mesh derp server {server_public_key:?}");
+
+            // receive detail loop
+            loop {
+                let (msg, conn_gen) = match self.recv_detail().await {
+                    Ok(res) => res,
+                    Err(e) => {
+                        tracing::warn!("recv error: {e:?}");
+                        tokio::time::sleep(redial_delay).await;
+                        break;
+                    }
+                };
+                if conn_gen != last_conn_gen {
+                    // TODO: In the current set up, once we know that this current
+                    // connection is a new connection, we must re-establish to the
+                    // server that we want to watch connection changes.
+                    // In the future, depending on how we handle multiple connections
+                    // for the same peer key in the derp server in the future, we may
+                    // be okay to just listen for future peer present
+                    // messages without re establishing this connection as a "watcher"
+                    break;
+                }
+                match msg {
+                    ReceivedMessage::PeerPresent(key) => {
+                        // ignore notifications about ourself
+                        if key == ignore_key {
+                            continue;
+                        }
+                        peers_present.insert(key.clone()).await;
+                        packet_forwarder_handler.add_packet_forwarder(key, self.clone())?;
```

Comment:
Is there a case where a peer is present, but no packet forwarder stored? or the other way around?

Reply:
No, we should be adding the packet forwarder in all cases.

ahh, are you asking if we are in trouble if `packet_forwarder_handle.add_packet_forwarder` returns in error?
If that happens, it means that the derp server has dropped its main actor channel, so the whole world has blown up.

Reply:
actually I was asking if we can avoid storing the duplicate keys in `peers_present` entirely 😅

Reply:
Ahhh I see.

Okay, so the `peers_present` map is entirely for logging purposes. Since it's specific for each client/remote derp server, it's logged here instead of in the main derp server.

I'd like to keep it in until we at least have metrics set up.